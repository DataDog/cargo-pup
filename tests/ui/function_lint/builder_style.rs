//@compile-flags: --crate-name test_builder_style
//@compile-flags: --crate-type lib

// This test verifies builder-style lints: "with_" methods returning `Self` are forbidden,
// while "set_" methods returning `&mut Self` are allowed.

pub struct WidgetBuilder {
    val: i32,
}

impl WidgetBuilder {
    // This should trigger the MustNotExist rule
    pub fn with_val(mut self, val: i32) -> Self { //~ ERROR: Function 'with_val' is forbidden by lint rule
        self.val = val;
        self
    }

    // This is the preferred style and should compile cleanly
    pub fn set_val(&mut self, val: i32) -> &mut Self {
        self.val = val;
        self
    }

    // Name matches the forbidden prefix but uses an allowed return type (&mut Self) – should be OK
    pub fn with_val_ref(&mut self, val: i32) -> &mut Self {
        self.val = val;
        self
    }

    // Opposite rule: name starts with "set_" but returns Self – should trigger error
    pub fn set_val_value(self, val: i32) -> Self { //~ ERROR: Function 'set_val_value' is forbidden by lint rule
        Self { val }
    }

    // Control method that matches neither rule
    pub fn touch(&self) {
        let _ = self.val;
    }
} 