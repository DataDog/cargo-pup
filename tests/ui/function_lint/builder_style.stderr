error: Function 'with_val' is forbidden by lint rule
  --> tests/ui/function_lint/builder_style.rs:13:5
   |
LL |     pub fn with_val(mut self, val: i32) -> Self {
   |     ^^^^^^^^^^^^^
   |
   = help: Remove this function to satisfy the architectural rule
   = note: Applied by cargo-pup rule 'builder_style_with_methods_forbidden'.
   = note: `#[deny(function_lint)]` on by default

error: Function 'set_val_value' is forbidden by lint rule
  --> tests/ui/function_lint/builder_style.rs:31:5
   |
LL |     pub fn set_val_value(self, val: i32) -> Self {
   |     ^^^^^^^^^^^^^^^^^^
   |
   = help: Remove this function to satisfy the architectural rule
   = note: Applied by cargo-pup rule 'builder_style_set_methods_forbid_self_value'.

error: aborting due to 2 previous errors

